### ISSUE TITLE
emulate set position and tilt by time driven down and up commands

### ISSUE THREAD

**Issue by bit72** [W00D00/home-assistant-elero#35]
Hello and big thanks to developer(s) of this component!

Since it is unfortunately not possible with elero transmitter stick to set position and tilt by HW (set_position, open_tilt, close_tilt, stop_tilt, set_tilt_position) would it maybe be possible to emulate this by time driven down and up commands? What i could experiece Elero MultiTel is doing so.

In case I explaned not detaild enough please let me know.
Thanks in advance!

**Comment by Schnabulation** [comment]
It's funny - I was thinking about this some days ago aswell. That would be very nice but has some drawbacks: It would only work from the top end position or bottom end position. Otherwise the time would not be accurate. But that would still be worth it. 

Maybe one could even code it so that if the roller is not in one of the end positions it would first go to the nearest end position and then move from there to the selected position. But don't forget: different roller lengths have different closing times.

**Comment by bit72** [comment]
Works best from top and end position... you are right. But theoretically - if only homeassiant is controling the cover - the actual position could be hold internally in this elero component as well. Sure this would not be perfect but it should work good enough. I have seen this in ESPHome. there is a time-based-cover (https://esphome.io/components/cover/time_based.html) which worked good for me on my shelly-based covers. 

For the beginning counting from top would be already great since even elero MultiTel cannot do more than that and at least I am somehow fine with this ;-)

**Comment by Schnabulation** [comment]
> if only homeassiant is controling the cover - the actual position could be hold internally in this elero component as well

That's a neat idea - but I demand this to be optional Because I also have wall remotes for the covers. :-D

If it helps I had a nice conversation with Elero some time ago and learned that the newer motors don't need to have the end stop learned in. They just roll down until the relay of the motor falls off which in turn means that the cover is in its end position. If the relay does not fall of it rolls for 3 minutes.

**Comment by u20p17** [comment]
Hi,
i am using this custom component for my roller shutters with elero motors and at the moment I am only getting 0%, 100% and 50% when it is moving. If I understand correctly this is due to the fact that the elero api does not send the actual position.

If I understand this issue correctly you were talking about a new feature which measures the total time the roller shutter needs from 0% to 100% and then it could calculate an aprox. position depending on the time which it needed? This would be very nice and exactly what I am looking for! Would be nice to have such a featue in the future! :)

Thanks to all developers!

**Comment by Schnabulation** [comment]
> If I understand correctly this is due to the fact that the elero api does not send the actual position

I think it's not the API that does not send the actual position - it's the motor itself that does not know the position. But yes, the idea would be that one could calibrate the runtime per shutter and then it would calculate the position based on the runtime. It would - however - only work from the top and the bottom position.

**Comment by u20p17** [comment]
that would be a very nice feature :D 

But why it would only work from top or bottom position? E.g.: The roller shutter is at 0%. Moving to 100% take the roller shutter 10s. Assuming the shutter moves always at constant speed it would take 2,5s for going to 25%. If I now stop the shutter and save the position (25%) and after a while want to go to 75% then I just have to drive another 5s in the same direction. So it should be possible to drive intermediate steps without always going to the top/lowest position. Of course after a reboot of HomeAssistant/custom component you have to go once to the top or lowest position again. Or do I overlook something?

**Comment by Schnabulation** [comment]
> Or do I overlook something?

Yes. :) I have a non-smart remote which I use to control the shutter without Home Assistant. HA does not know the position if I adjust it using the remote.

**Comment by u20p17** [comment]
Okay, didn't thought about that. I mainly use NodeRed and HomeAssistant to control my roller shutters... But I have a MultiTel2 remote control and did some tests now: If I press move downwards on the remote control the elero state in HomeAssistant also changes from 'top position stop' to 'moving down'. So it would also be possible to detect manual roller shutter movements (via the remote control) and therefore it should be possible to move from example 25% to 75% (you just would have to start the timer when the elero state changes...) without opening or closing it completely... maybe you have a different constellation at your home where this is not valid?

**Comment by u20p17** [comment]
I saw that @dfigus also commited to this repository the last days - maybe he could also help us with this topic :)

**Comment by W00D00** [comment]
Hi Guys,

unfortunately, the Elero system does not support the state info. So we cannot get any info about the real position of the shutters just what I have developed. Plus, the Elero Transmitter Stick is buggy.  I would be happy that I can serve you with this feature however, as I see now it is not possible. The time-based solution would be not as accurate as you expected. So, I do now want to spend time on this. However, if you describe your idea better in written form to me, I can fork a dev branch where we can try out something without any risk of this working main branch. Hm, so...

Thanks,
István

**Comment by Schnabulation** [comment]
Hi István

Thank you for taking over this project - really appreciate further development.

I am aware that Elero does not support the state info appart from "open", "closed" and "vent/tilt". However our idea is based on the constant movement speed of the motors. Meaning: it always takes the same time from fully open to fully closed and the other way around.

So knowing this we could calibrate each shutter by measuring the total time it needs to travel and then based on this, calculate the travel time for each shutter position.

Ex: the shutter needs 30 seconds from fully open to fully closed. If we now want to have the shutter go to 20% closed we move it 6 seconds (30 / 100 * 20). If the shutter is in an unknown state, we first move it to the very top and then move it down 6 seconds.

**Comment by u20p17** [comment]
any news on this topic? would be nice if @W00D00 has some time for the dev branch which we could at least try and see if the idea with the 'position calculation' based on the constant movement speed of the motors is good enough or not. :)

**Comment by W00D00** [comment]
Hi All,

unfortunately, I do not have time for this development. Next to this, as I know, we cannot know so much about the running time of the motors, because we do not know exactly when the command was sent out to the motor. After all, the HA system does not provide a callback to this. If someone knows something special, please share the knowledge and experience to verify the idea. 

Thanks,
István

**Comment by u20p17** [comment]
Hi @W00D00 ,
I did a bit of research and found out that the knx integration does something similiar: 

https://github.com/XKNX/xknx/blob/main/xknx/devices/travelcalculator.py

I think that we could use the 'elero state' as the trigger for the command send from HA, because when I send a command I immediately can see the state changing.

What do you think?

**Comment by ma-ku** [comment]
Hi all, this is a bit stale here but I am the maintainer of the Elero-Plugin for HomeBridge https://github.com/ma-ku/homebridge-elero-stick and I do support percentages for the openings. It is accomplished through (a) setting the time in the settings for a shutter and (b) polling the motor state during activities. 

I am planning to move from HomeBridge to HA and would be happy to help out getting this plugin on par with the Homebridge version.

One more comment regarding the vent/tilt position; technically it seems as if you can set these positions in the motors and then you should be able to move to these programmed positions. Since I do not need this, I have never tested it. Also is the Elero logic used across many different devices (including lights) so not every functionality is necessarily targeted to shutters and comparable devices.

**Comment by Schnabulation** [comment]
Hi @ma-ku 
Great to hear from you! I am a long time user of the Elero integration for Home Assistant and still heavily depend on it. It would be amazing to get help from you.

Regarding the vent/tilt position: yes, this is correct. You can set these positions directly in the motors with the remote. I have done this for my home and have set the vent position to be all the way down and then having the slats horizontal. I can therefore directly set this position from Home Assistant and use it quite often.

If you need me to test anything I'm happy to help!

**Comment by ma-ku** [comment]
Hi @Schnabulation, great to hear. I will need to see how fast I can do this. Probably we can only tackle this by switching to the async model as we will need to run quite a few checks in the background to mimic the percentages and so on. Since I have about eight Elero motors in my house, I am a direct beneficiary of that effort. Since I do Python only once in a while, I will need some time to wrap my head around this. Ideally I can just port the logic from Node.JS to Python and add these additional features like the ventilation position. I had this already prepared in the HomeBridge plugin but the question is, how this should be represented in HomeKit for example (would that be an additional push-button)?

I will keep you in the loop so we can test together.

**Comment by Schnabulation** [comment]
I'm happy to help - if I can. I have some programming background but am a total Python noob. So not much I can help there I guess. But I'm very familiar with Home Assistant and can most certainly beta-test features if you want me to.

**Comment by thomaskonrad** [comment]
The [Becker integration](https://github.com/RainerStaude/hass-becker-component-plus-pybecker) does exactly this by using [TravelCalculator](https://github.com/RainerStaude/hass-becker-component-plus-pybecker/blob/main/travelcalculator.py) from the [XKNX project](https://github.com/XKNX/xknx/blob/5d2751658b62c4f9f709870fa85e1d18cdb4e37d/xknx/devices/travelcalculator.py#L27).
